src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java

```diff
@@ -119,6 +130,42 @@ private void process(final DuplicateSet set) {
         }
 
         final UmiGraph umiGraph = new UmiGraph(set, umiTag, inferredUmiTag, allowMissingUmis);
-        nextSetsIterator = umiGraph.joinUmisIntoDuplicateSets(maxEditDistanceToJoin).iterator();
+
+        List<DuplicateSet> duplicateSets = umiGraph.joinUmisIntoDuplicateSets(maxEditDistanceToJoin);
+
+        // Collect statistics on numbers of observed and inferred UMIs
+        // and total numbers of observed and inferred UMIs
+        for (DuplicateSet ds : duplicateSets) {
+            List<SAMRecord> records = ds.getRecords();
+            SAMRecord representativeRead = ds.getRepresentative();
+            String inferredUmi = representativeRead.getStringAttribute(inferredUmiTag);
+
+            for (SAMRecord rec : records) {
+                String currentUmi = rec.getStringAttribute(umiTag);
+
+                if (currentUmi != null) {
+                    // All UMIs should be the same length
+                    if (!haveWeSeenFirstRead) {
+                        metrics.UMI_LENGTH = currentUmi.length();
+                        haveWeSeenFirstRead = true;
+                    } else {
+                        if (metrics.UMI_LENGTH != currentUmi.length()) {
+                            throw new PicardException("UMIs of differing lengths were found.");
```

Comment:
Should there be a TODO to allow UMI's of different length? perhaps UMI_LENGTH should be a double to mean "average" umi length even if we do not allow that right now?